backstory/src/demo/java/com/google/sps/servlets/PerspectiveServlet.java

```diff
@@ -39,11 +46,90 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     response.setContentType("application/json;");
     Gson gson = new Gson();
 
-    String[] cannedOutput = new String[0];
+    String json = request.getReader().readLine();
 
-    // write canned output to servlet
-    String json = gson.toJson(cannedOutput);
+    
+    // get the text to be analyzedfrom the JSON & handle error if it cannot be converted
+    String text = null;
+    try {
+      JSONObject jsonObject = new JSONObject(json);
+      text = jsonObject.getString("text");
+    } catch (JSONException exception) {
+      String errorMessage = "Could not convert text sent to server from JSON.";
 
-    response.getWriter().println(json);
+      handleError(response, HttpServletResponse.SC_BAD_REQUEST, errorMessage);
+      return;
+    }
+
+    // check that text is valid (not null or empty)
+    if (text == null || text.equals("")) {
+      handleError(response, HttpServletResponse.SC_BAD_REQUEST, "Text input was null or empty");
+      return;
+    }
+
+    // as we want to display the values as well as the decision in our demo
+    // we need to access internal code that the MVP would not
+    PerspectiveValues values = null;    
+    try {
+      PerspectiveAPIFactory factory = new PerspectiveAPIFactoryImpl();
+      PerspectiveAPI api = factory.newInstance();
+      PerspectiveAPIClient apiClient = new PerspectiveAPIClient(api);
+      values = apiClient.analyze(Arrays.asList(PerspectiveStoryAnalysisManager.REQUESTED_ATTRIBUTES), text);
+    } catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException | InvocationTargetException exception) {
+      // if any errors were thrown for looking for api key, send that it was not retrieved back to page
+      handleError(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Could not retrieve the Perspective API key.");
+      return;
+    }
+
+    // the only code needed for the data pipeline in the MVP is 
+    // the construction of a PerspectiveManager object and 
+    // a call to the generateDecision() method of PerspectiveManager
+    
+    // create a StoryAnalysisManager
+    StoryAnalysisManager manager;
+    try {
+      manager = new PerspectiveStoryAnalysisManager();
+    } catch (APINotAvailableException exception) {   
+      handleError(response, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, exception.toString());   
+      return;
+    }
+
+    // generate a StoryDecision with that manager
+    StoryDecision storyDecision;
+    try {
+      storyDecision = manager.generateDecision(text);
+    } catch (NoAppropriateStoryException e) {
+      storyDecision = null;
+    }
+
+    // if storyDecision is null, then it means the decision is not appropriate (should be false)
+    Boolean isAppropriateStory = (storyDecision != null);
+
+    // objects to write back on the respones
+    Object[] objectsToWrite = { isAppropriateStory, values.getAttributeTypesToScores() };
```

Comment:
I think we discussed using a map instead?

HashMap<String, Object> responseBody = new HashMap<>();
responseBody.put("isAppropriate", isAppropriateStory);
responseBody.put("attributeTypesToScores, values.getAttributeTypesToScores());
response.getWriter().println(gson.toJson(responseBody));

Reply:
I agree. I'll change this.

Reply:
Done